components/script/dom/intersectionobserver.rs

```diff
@@ -584,6 +584,142 @@ impl IntersectionObserver {
         )
     }
 
+    /// <https://w3c.github.io/IntersectionObserver/#calculate-visibility-algo>
+    pub(crate) fn calculate_visibility(&self, document: &Document, target: &Element) -> bool {
+        // Step 1
+        // > If the observer’s trackVisibility attribute is false, return false.
+        if !self.track_visibility.get() {
+            return false;
+        }
+
+        // Step 2
+        // > If the target has an effective transformation matrix other than a 2D translation or proportional 2D upscaling, return false.
+        let effective_transformation_matrix =
+            self.calculate_effective_transformation_matrix(document, target);
+
+        if !effective_transformation_matrix.is_2d() {
+            // TODO proportional 2d upscaling?
+            return false;
+        }
+
+        // Step 3
+        // > If the target, or any element in its containing block chain, has an effective opacity other than 100%, return false.
+        if target.style().unwrap().clone_opacity() != 1.0 {
+            return false;
+        } // TODO containing block chain
+
+        // Step 4
+        // > If the target, or any element in its containing block chain, has any filters applied, return false.
+        if !target.style().unwrap().clone_filter().0.is_empty() {
+            return false;
+        }
+
+        // Step 5
+        // > If the implementation cannot guarantee that the target is completely unoccluded by other page content, return false.
+        // TODO Not sure how to do this bit
+
+        // Step 6
+        // > Return true.
+        true
+    }
+
+    /// <https://w3c.github.io/IntersectionObserver/#calculate-effective-transformation-matrix>
+    pub(crate) fn calculate_effective_transformation_matrix(
+        &self,
+        document: &Document,
+        target: &Element,
+    ) -> Transform3D<f32, style_traits::CSSPixel, style_traits::CSSPixel> {
+        // Step 1
+        // > Let matrix be the serialization of the identity transform function.
+        let mut matrix: Transform3D<f32, style_traits::CSSPixel, style_traits::CSSPixel> =
+            Transform3D::from_untyped(
+                &target
+                    .style()
+                    .unwrap()
+                    .clone_transform()
+                    .to_transform_3d_matrix(None)
+                    .unwrap()
+                    .0,
+            );
+
+        // Step 2
+        // > Let container be the target.
+        #[allow(unused_mut)] // TODO remove
+        let mut container = target;
+
+        // Step 3
+        // > While container is not the intersection root:
+        #[allow(clippy::never_loop)] // TODO remove
+        while if self.root_is_implicit_root() {
+            *container.owner_document() != *document
```

Comment:
I don't think this condition is correct for nested browsing contexts; the document that is passed in is not the top level browsing context's document node per https://w3c.github.io/IntersectionObserver/#intersectionobserver-intersection-root.